### PR TITLE
Add lstm with recurrent projection layer operator

### LINKED ISSUE #7547: Implement  LSTM with projection layer (LSTMP)

### PR DESCRIPTION
**Pull request by kuke**
Fix #7547

### CODE REVIEW

File: paddle/operators/lstmp_op.h [hunk 1]
```diff
@@ -0,0 +1,489 @@
...
+        // to reorder.
+
+        Tensor ordered_h0;
+        ordered_proj0->mutable_data<T>(ctx.GetPlace());
+        ReorderInitState<DeviceContext, T>(device_ctx, *hidden_t0, order,
+                                           &ordered_h0, true);
+        math::matmul<DeviceContext, T>(device_ctx, ordered_h0, false,
+                                       *proj_weight, false, static_cast<T>(1.0),
+                                       ordered_proj0, static_cast<T>(0.0));
+        if (share_cell_act) {
```
Comment (pkuyym) [inline paddle/operators/lstmp_op.h]: Please consider to use string type.

File: paddle/operators/lstmp_op.cc [hunk 2]
```diff
@@ -0,0 +1,319 @@
...
+namespace paddle {
+namespace operators {
+
+class LSTMPOp : public framework::OperatorWithKernel {
+ public:
+  using framework::OperatorWithKernel::OperatorWithKernel;
+
+  void InferShape(framework::InferShapeContext* ctx) const override {
+    PADDLE_ENFORCE(ctx->HasInput("Input"),
+                   "Input(Input) of LSTMP should not be null.");
```
[the remark this answers is not in the document]
Comment (kuke) [reply paddle/operators/lstmp_op.cc]: Done

File: paddle/operators/lstmp_op.cc [hunk 3]
```diff
@@ -0,0 +1,319 @@
...
+                      "The rank of Input(ProjWeight) should be 2.");
+    PADDLE_ENFORCE_EQ(proj_dims[0], frame_size,
+                      "The first dimension of Input(ProjWeight) "
+                      "should be %d.",
+                      frame_size);
+
+    if (ctx->HasInput("H0")) {
+      PADDLE_ENFORCE(ctx->HasInput("C0"),
+                     "Input(C0) and Input(H0) of LSTMP should not "
+                     "be null at the same time.");
```
[the remark this answers is not in the document]
Comment (kuke) [reply paddle/operators/lstmp_op.cc]: Done

File: paddle/operators/lstmp_op.h (hunk 1 above)
Comment (kuke) [reply paddle/operators/lstmp_op.h]: Done

**Pull request merged**